logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java

```diff
@@ -76,31 +76,53 @@ public WorkerLoop(
     public void run() {
         try {
             boolean isShutdown = false;
+            boolean isNackBatch = false;
             do {
                 isShutdown = isShutdown || shutdownRequested.get();
                 final QueueBatch batch = readClient.readBatch();
                 final boolean isFlush = flushRequested.compareAndSet(true, false);
                 if (batch.filteredSize() > 0 || isFlush) {
                     consumedCounter.add(batch.filteredSize());
-                    execution.compute(batch, isFlush, false);
-                    filteredCounter.add(batch.filteredSize());
-                    readClient.closeBatch(batch);
-
-                    if (isFlush) {
-                        flushing.set(false);
+                    isNackBatch = abortableCompute(batch, isFlush, false);
+                    if (!isNackBatch) {
+                        filteredCounter.add(batch.filteredSize());
+                        readClient.closeBatch(batch);
+
+                        if (isFlush) {
+                            flushing.set(false);
+                        }
                     }
                 }
-            } while (!isShutdown || isDraining());
-            //we are shutting down, queue is drained if it was required, now  perform a final flush.
-            //for this we need to create a new empty batch to contain the final flushed events
-            final QueueBatch batch = readClient.newBatch();
-            execution.compute(batch, true, true);
-            readClient.closeBatch(batch);
+            } while ((!isShutdown || isDraining()) && !isNackBatch);
+
+            if (!isNackBatch) {
+                //we are shutting down, queue is drained if it was required, now  perform a final flush.
+                //for this we need to create a new empty batch to contain the final flushed events
+                final QueueBatch batch = readClient.newBatch();
+                abortableCompute(batch, true, true);
+                readClient.closeBatch(batch);
+            }
         } catch (final Exception ex) {
             throw new IllegalStateException(ex);
         }
     }
 
+    private boolean abortableCompute(QueueBatch batch, boolean flush, boolean shutdown) {
+        boolean isNackBatch = false;
+        try {
+            execution.compute(batch, flush, shutdown);
+        } catch (Exception ex) {
+            if (ex instanceof AbortedBatchException) {
+                isNackBatch = true;
```

Comment:
Is there a reason we don't only `catch (AbortedBatchException ex)` here? If so, can we document the reason?

Reply:
The motivation is the compiler complains about an error that honestly is not so clear to me why.
The error is:
```
logstash-core/src/main/java/org/logstash/execution/WorkerLoop.java:114: error: exception AbortedBatchException is never thrown in body of corresponding try statement
        } catch (AbortedBatchException aex) {
          ^
1 error
```

Because the `compile` method signature doesn't declare that it could throw `AbortedBatchException`, but it also neither declare  `Exception`, and doesn't complain about that.
`AbortedBatchException` subclasses `Exception`.

https://github.com/elastic/logstash/blob/c99a7b2f46f7fbf64ae64a45d2f546f50a7d8523/logstash-core/src/main/java/org/logstash/config/ir/CompiledPipeline.java#L361

Reply:
We could (and maybe it's more clean) to add to `CompiledPipeline` signature the ability throw `AbortedBatchException`, because it's more explicit and it's effectively something that could happen and the caller must be prepared to handle it.